api/src/lib/github/repository.go

```diff
@@ -0,0 +1,72 @@
+package github
+
+import (
+	"strconv"
+	"io/ioutil"
+	"encoding/json"
+	"net/http"
+	"porter/api/src/config"
+	"sync"
+	"fmt"
+)
+
+type RepositoryResponse struct {
+	Url string `json:"url"`
+	Name string `json:"name"`
+	Html_url string `json:"html_url"`
+}
+
+const repositoryPages = 20
+
+type Repository struct {
+	Name         string
+	ExposedPorts []string
+	Url 		string
+	Html_url string
+}
+
+
+func (rp Repository) MarshalBinary() ([]byte, error) {
+	return json.Marshal(&rp)
+}
+
+func (rp *Repository) UnMarshalBinary(data []byte) (error) {
+	return json.Unmarshal(data, &rp)
+}
+
+func FetchRepository(page int,ch chan []RepositoryResponse){
+	var repositoryResponse []RepositoryResponse
+	resp,_ := http.Get("https://api.github.com/orgs/fiverr/repos?access_token="+ config.GetValue("githubToken") +"&limit=100&sort=pushed&page=" + strconv.Itoa(page))
+	body, err := ioutil.ReadAll(resp.Body)
+	if err != nil {
+		fmt.Println(err)
+	}
+
+	json.Unmarshal(body, &repositoryResponse)
+	ch <- repositoryResponse
+}
+
+func FetchAllRepositories() ([]RepositoryResponse) {
+	repositories := make([]RepositoryResponse, 0)
+	repositoriesResponseCh := make(chan []RepositoryResponse)
+	var wg sync.WaitGroup
+
+	for i:= 1; i <= repositoryPages ; i++  {
+		go FetchRepository(i, repositoriesResponseCh)
+	}
+
+	go func(ch chan []RepositoryResponse) {
+		wg.Add(repositoryPages)
+		wg.Wait()
+		close(ch)
+	}(repositoriesResponseCh)
```

Comment:
@fiverrben is it a good practice?

Reply:
In order for your code to be more "generic" I think the better way of doing this is...

``` 
for i:= 1; i <= repositoryPages ; i++  {
        wg.Add(1)
	go FetchRepository(i, repositoriesResponseCh, &wg)
}
wg.Wait()
```

and then `FetchRepository` looks something like:
```
func FetchRepository(page int,ch chan []RepositoryResponse, wg *sync.WaitGroup){
    defer wg.Done()
    ...
}
```

![image](https://user-images.githubusercontent.com/11057911/39959779-fbcf9fb6-55e4-11e8-95fe-deb23ebe595a.png)


See a good article about usage: https://nathanleclaire.com/blog/2014/02/15/how-to-wait-for-all-goroutines-to-finish-executing-before-continuing/

Reply:
I saw this pattern in https://blog.golang.org/pipelines, 
if I'll change it there will be a dead lock